File: genkit-tools/cli/src/mcp/usage.ts
/**
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { record } from '@genkit-ai/tools-common/utils';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp';
import { ContentBlock } from '@modelcontextprotocol/sdk/types';
import z from 'zod';
import { McpRunToolEvent } from './analytics.js';

import { GENKIT_CONTEXT as GoContext } from '../commands/init-ai-tools/context/go.js';
import { GENKIT_CONTEXT as JsContext } from '../commands/init-ai-tools/context/nodejs.js';

export async function defineUsageGuideTool(server: McpServer) {
  server.registerTool(
    'get_usage_guide',
    {
      title: 'Genkit Instructions',
      description:
        'Use this tool to look up the Genkit usage guide before implementing any AI feature',
      inputSchema: {
        language: z
          .enum(['js', 'go'])
          .describe('which language this usage guide is for')
          .default('js')
          .optional(),
      },
    },
    async ({ language }) => {
      await record(new McpRunToolEvent('get_usage_guide'));

      const content = [] as ContentBlock[];
      if (!language) {
        language = 'js';
      }
      let text = JsContext;
      if (language === 'go') {
        text = GoContext;
      }
      content.push({
        type: 'text',
        text,
      });

      return { content };
    }
  );
}

File: java/ai/src/main/java/com/google/genkit/ai/Usage.java
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.google.genkit.ai;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Usage represents token usage statistics from a model response.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class Usage {

  @JsonProperty("inputTokens")
  private Integer inputTokens;

  @JsonProperty("outputTokens")
  private Integer outputTokens;

  @JsonProperty("totalTokens")
  private Integer totalTokens;

  @JsonProperty("inputCharacters")
  private Integer inputCharacters;

  @JsonProperty("outputCharacters")
  private Integer outputCharacters;

  @JsonProperty("inputImages")
  private Integer inputImages;

  @JsonProperty("outputImages")
  private Integer outputImages;

  @JsonProperty("inputAudioFiles")
  private Integer inputAudioFiles;

  @JsonProperty("outputAudioFiles")
  private Integer outputAudioFiles;

  @JsonProperty("inputVideoFiles")
  private Integer inputVideoFiles;

  @JsonProperty("outputVideoFiles")
  private Integer outputVideoFiles;

  @JsonProperty("thoughtsTokens")
  private Integer thoughtsTokens;

  @JsonProperty("cachedContentTokens")
  private Integer cachedContentTokens;

  /**
   * Default constructor.
   */
  public Usage() {
  }

  /**
   * Creates a Usage with token counts.
   *
   * @param inputTokens
   *            number of input tokens
   * @param outputTokens
   *            number of output tokens
   * @param totalTokens
   *            total number of tokens
   */
  public Usage(Integer inputTokens, Integer outputTokens, Integer totalTokens) {
    this.inputTokens = inputTokens;
    this.outputTokens = outputTokens;
    this.totalTokens = totalTokens;
  }

  // Getters and setters

  public Integer getInputTokens() {
    return inputTokens;
  }

  public void setInputTokens(Integer inputTokens) {
    this.inputTokens = inputTokens;
  }

  public Integer getOutputTokens() {
    return outputTokens;
  }

  public void setOutputTokens(Integer outputTokens) {
    this.outputTokens = outputTokens;
  }

  public Integer getTotalTokens() {
    return totalTokens;
  }

  public void setTotalTokens(Integer totalTokens) {
    this.totalTokens = totalTokens;
  }

  public Integer getInputCharacters() {
    return inputCharacters;
  }

  public void setInputCharacters(Integer inputCharacters) {
    this.inputCharacters = inputCharacters;
  }

  public Integer getOutputCharacters() {
    return outputCharacters;
  }

  public void setOutputCharacters(Integer outputCharacters) {
    this.outputCharacters = outputCharacters;
  }

  public Integer getInputImages() {
    return inputImages;
  }

  public void setInputImages(Integer inputImages) {
    this.inputImages = inputImages;
  }

  public Integer getOutputImages() {
    return outputImages;
  }

  public void setOutputImages(Integer outputImages) {
    this.outputImages = outputImages;
  }

  public Integer getInputAudioFiles() {
    return inputAudioFiles;
  }

  public void setInputAudioFiles(Integer inputAudioFiles) {
    this.inputAudioFiles = inputAudioFiles;
  }

  public Integer getOutputAudioFiles() {
    return outputAudioFiles;
  }

  public void setOutputAudioFiles(Integer outputAudioFiles) {
    this.outputAudioFiles = outputAudioFiles;
  }

  public Integer getInputVideoFiles() {
    return inputVideoFiles;
  }

  public void setInputVideoFiles(Integer inputVideoFiles) {
    this.inputVideoFiles = inputVideoFiles;
  }

  public Integer getOutputVideoFiles() {
    return outputVideoFiles;
  }

  public void setOutputVideoFiles(Integer outputVideoFiles) {
    this.outputVideoFiles = outputVideoFiles;
  }

  public Integer getThoughtsTokens() {
    return thoughtsTokens;
  }

  public void setThoughtsTokens(Integer thoughtsTokens) {
    this.thoughtsTokens = thoughtsTokens;
  }

  public Integer getCachedContentTokens() {
    return cachedContentTokens;
  }

  public void setCachedContentTokens(Integer cachedContentTokens) {
    this.cachedContentTokens = cachedContentTokens;
  }
}
